src/lib/metrics.ts

```diff
@@ -0,0 +1,142 @@
+  endTimeMs = 0;
+
+  public getValue(): MetricValue {
+    if (this.startTimeMs != 0 && this.endTimeMs != 0) {
+      return this.endTimeMs - this.startTimeMs;
+    } else {
+      return undefined;
+    }
+  }
+
+  public start() {
+    if (this.startTimeMs === 0) {
+      this.startTimeMs = Date.now();
+    } else {
+      logImproperUse(
+        'Invalid Timer use: start() called when timer already stopped',
+      );
+    }
+  }
+
+  public stop() {
+    if (this.endTimeMs === 0) {
+      this.endTimeMs = Date.now();
+      debug(
+        `Timer stopped at ${
+          this.endTimeMs
+        }. Elapsed time is ${this.getValue()}`,
+      );
+    } else {
+      logImproperUse(
+        'Invalid Timer use: stop() called when timer already stopped',
+      );
+    }
+  }
+}
+
+export class SyntheticMetricInstance extends MetricInstance {
+  private value = 0;
+
+  public setValue(value: number) {
+    this.value = value;
+  }
+
+  public getValue(): number {
+    return this.value;
+  }
+}
+
+export abstract class Metric {
+  public name: string;
+  public context: string;
+  public metricType: MetricType;
+  protected instances: Array<MetricInstance> = [];
+
+  public clear() {
+    this.instances = [];
+  }
+
+  public getValues(): number[] {
+    return this.instances.map((mi) => mi.getValue() || 0);
+  }
+
+  public getTotal(): number {
+    const sumMetricValues = (accum: number, current: MetricInstance) => {
+      const currentTimerMs = current.getValue() || 0;
+      return (accum = accum + currentTimerMs);
+    };
+
+    const total = this.instances.reduce(sumMetricValues, 0);
+    return total;
+  }
+
+  public constructor(name: string, metricType: MetricType, context: string) {
+    this.name = name;
+    this.metricType = metricType;
+    this.context = context;
+  }
+}
+
+export class TimerMetric extends Metric {
+  public createInstance(): TimerMetricInstance {
+    const t = new TimerMetricInstance();
+    this.instances.push(t);
+    return t;
+  }
+}
+
+export class SyntheticMetric extends Metric {
+  public createInstance(): SyntheticMetricInstance {
+    const sm = new SyntheticMetricInstance();
+    this.instances.push(sm);
+    return sm;
+  }
+}
+
+export function logImproperUse(msg: string) {
+  debug(msg);
+}
+
+export class MetricsCollector {
+  public static NETWORK_TIME: TimerMetric = new TimerMetric(
+    'network-time',
+    'Timer',
+    'the context',
+  );
+  public static CPU_TIME: SyntheticMetric = new SyntheticMetric(
+    'cpu-time',
+    'Synthetic',
+    'the context',
+  );
+
+  public static getAllMetrics(): any[] {
+    const res: any[] = [];
+    const metrics: Metric[] = [
+      MetricsCollector.NETWORK_TIME,
+      MetricsCollector.CPU_TIME,
+    ];
+    metrics.forEach((m) => {
+      const nextEntry = {
+        name: m.name,
+        type: m.metricType,
+        values: m.getValues(),
+        total: m.getTotal(),
+      };
+      res.push(nextEntry);
+    });
+    return res;
+  }
+}
```

Comment:
You could probably map these:

```suggestion
  public static getAllMetrics(): any[] {
    const metrics: Metric[] = [
      MetricsCollector.NETWORK_TIME,
      MetricsCollector.CPU_TIME,
    ];
    return metrics.map(m => ({
      name: m.name,
      type: m.metricType,
      values: m.getValues(),
      total: m.getTotal(),
    }));
  }
}
```

Reply:
Thanks